packages/core/codeframe/src/codeframe.js

```diff
@@ -116,102 +125,166 @@ export default function codeFrame(
       : endLine;
   let endLineString = endLine.toString(10);
 
-  let resultLines = [];
+  // Split input into lines and highlight syntax
   let lines = code.split(NEWLINE);
   let syntaxHighlightedLines = (opts.syntaxHighlighting
     ? highlightSyntax(code, opts.language)
     : code
   )
     .replace(TAB_REPLACE_REGEX, TAB_REPLACEMENT)
     .split(NEWLINE);
-  for (let i = startLine; i < lines.length; i++) {
-    if (i > endLine) break;
 
-    let originalLine: string = lines[i];
-    let syntaxHighlightedLine = syntaxHighlightedLines[i];
+  // Loop over all lines and create codeframe
+  let resultLines = [];
+  for (
+    let currentLineIndex = startLine;
+    currentLineIndex < syntaxHighlightedLines.length;
+    currentLineIndex++
+  ) {
+    if (currentLineIndex > endLine) break;
+    if (currentLineIndex > syntaxHighlightedLines.length - 1) break;
+
+    // Find highlights that need to get rendered on the current line
+    let lineHighlights = highlights
+      .filter(
+        highlight =>
+          highlight.start.line <= currentLineIndex &&
+          highlight.end.line >= currentLineIndex,
+      )
+      .sort(
+        (a, b) =>
+          (a.start.line < currentLineIndex ? 0 : a.start.column) -
+          (b.start.line < currentLineIndex ? 0 : b.start.column),
+      );
 
-    let foundHighlights = highlights.filter(
-      highlight => highlight.start.line <= i && highlight.end.line >= i,
-    );
+    // Check if this line has a full line highlight
+    let isWholeLine =
+      lineHighlights.length &&
+      !!lineHighlights.find(
+        h => h.start.line < currentLineIndex && h.end.line > currentLineIndex,
+      );
+
+    let lineLengthLimit = 1000;
+    if (opts.terminalWidth && opts.terminalWidth > endLineString.length + 7) {
+      lineLengthLimit = opts.terminalWidth - (endLineString.length + 5);
+    }
+
+    // Split the line into line parts that will fit the provided terminal width
+    let colOffset = 0;
+    let lineEndCol = lineLengthLimit;
+    let syntaxHighlightedLine = syntaxHighlightedLines[currentLineIndex];
+    if (stringWidth(syntaxHighlightedLine) > lineLengthLimit) {
+      if (lineHighlights.length > 0) {
+        if (lineHighlights[0].start.line === currentLineIndex) {
+          colOffset = lineHighlights[0].start.column - 5;
```

Comment:
Are you subtracting 5 because of the gutter on left? But that can be wider than 5 as well

Reply:
It's just to add some padding so you have some context

Reply:
https://github.com/parcel-bundler/parcel/blob/0f259215447f9310da59e1d84830f2dc5bda5ea9/packages/core/codeframe/src/codeframe.js#L169 this is the line which handles the gutter on the left

Reply:
I see

![Bildschirmfoto 2020-05-13 um 13 10 41](https://user-images.githubusercontent.com/4586894/81805463-2d024300-951b-11ea-8007-0217074ff3a6.png)

(Though this size is impractical anyway)